tools/gencat/gencat.1
.\"	$OpenBSD: gencat.1,v 1.8 1999/06/05 01:21:26 aaron Exp $
.\"
.\" Copyright (c) 1997 Ken Stailey
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	$Id: gencat.1,v 1.2 2000/08/14 16:26:08 amai Exp $
.\"
.Dd June 11, 1997
.Dt GENCAT 1
.Os
.Sh NAME
.Nm gencat
.Nd NLS catalog compiler
.Sh SYNOPSIS
.Nm gencat
.Ar output-file
.Ar input-file Op Ar input-file...
.Sh DESCRIPTION
The
.Nm
utility merges the text NLS (National Language Support) in
.Ar input-file
into a formatted message catalog file
.Ar output-file .
The file
.Ar output-file
will be created if it does not already exist.  If
.Ar output-file
does exist, its messages will be included in the new
.Ar output-file .
If set and message numbers collide, the new message text defined in
.Ar input-file
will replace the old message text currently contained in
.Ar output-file .
.Sh INPUT FILES
The format of a message text source file is defined below.  Note that
the fields of a message text source line are separated by a single space
character: any other space characters are considered to be part of the
field contents.
.Pp
.Bl -tag -width
.It Li $set Ar n comment
This line specifies the set identifier of the following messages until
the next
.Li $set
or end-of-file appears.  The argument
.Ar n
is the set identifier which is defined as a number in the range
[1, (NL_SETMAX)].  Set identifiers must occur in ascending order within
a single source file, but need not be contiguous.  Any string following
a space following the set identifier is treated as a comment.  If no
.Li $set
directive  is specified in a given source file, all messages will
be located in the default message set NL_SETD.
.It Li $del Ar n comment
This line deletes messages from set
.Ar n
from a message catalog.  The
.Ar n
specifies a set number.  Any string following a space following the set
number is treated as a comment.
.It Li $ Ar comment
A line beginning with
.Li $
followed by a space is treated as a comment.
.It Ar m message-text
A message line consists of a message identifier
.Ar m
in the range [1, (NL_MSGMAX)].  The
.Ar message-text
is stored in the message catalog with the set identifier specified by
the last
.Li $set
directive, and the message identifier
.Ar m .
If the
.Ar message-text
is empty, and there is a space character following the message identifier,
an empty string is stored in the message catalog.  If the
.Ar message-text
is empty, and if there is no space character following the message
identifier, then the existing message in the current set with the
specified message identifier is deleted from the catalog.  Message
identifiers must be in ascending order within a single set, but
need not be contiguous.  The
.Ar message-text
length must be in the range [0, (NL_TEXTMAX)].
.It Li $quote Ar c
This line specifies an optional quote character
.Ar c
which can be used to surround
.Ar message-text
so that trailing space or empty messages are visible in message
source files.  By default, or if an empty
.Li $quote
directive is specified, no quoting of
.Ar message-text
will be recognized.
.El
.Pp
Empty lines in message source files are ignored.  The effect of lines
beginning with any character other than those described above is
undefined.
.Pp
Text strings can contain the following special characters and escape
sequences.  In addition, if a quote character is defined, it may be
escaped as well to embed a literal quote character.
.Pp
.Bl -tag -width Ds -offset indent
.It Li \en
line feed
.It Li \et
horizontal tab
.It Li \ev
vertical tab
.It Li \eb
backspace
.It Li \er
carriage return
.It Li \ef
form feed
.It Li \e\e
backslash
.It Li \eooo
octal number in the range [000, 377]
.El
.Pp
A backslash character immediately before the end of the line in a file
is used to continue the line onto the next line, e.g.:
.Pp
.Dl 1 This line is continued \e
.Dl on this line.
.Pp
If the character following the backslash is not one of those specified,
the backslash is ignored.
.Pp
The
.Nm
utility exits 0 on success, or >0 if an error occurs.
.Sh SEE ALSO
.Xr catclose 3 ,
.Xr catgets 3 ,
.Xr catopen 3
.Sh STANDARDS
The
.Nm
utility is compliant with the
.St -xpg4
standard.
.Sh AUTHOR
This manual page by
.An Ken Stailey
updated and revised by
.An Terry Lambert .
.Sh BUGS
A message catalog file created from a blank input file can not be revised;
it must be deleted and recreated.


ChangeLog
		[$Id: ChangeLog,v 1.34 2000/11/27 18:13:42 arnd Exp $]
 
2000/03/14 LIBREVISION 0.01
	Initial version (arnd)
2000/03/22
	added fixcrcr utility and missing Makefiles in libext/uni (veit)

2000/03/24 LIBREVISION 0.02
	big change: (arnd)
	beautified fixcrcr.c, more docs, added syslog/drand/valloc/dlfcn,
	more headers, added some functions from EMX stat.c, added some
	i387 libm asm funcs, added a math test vector
2000/03/29 
	additions: (arnd)
	signal,resource,time header files and man pages
	Updated CodeRules
	Added Id lines to various files
	Support for itimer and daemon
	Misc fixes

2000/04/03 LIBREVISION 0.03
	additions: (arnd)
	more BSD man2 docs, signal(3) docs
	some missing headers (limits, signal, cdefs, itimer, resource)
	signalx.c was missing in CVS
	popen.c
	fixed a few Makefiles
	more CVSIds in files

2000/04/17 LIBREVISION 0.04 (0.04a)
	additions: (arnd)
	more docs, several sys headers from BSD
	regex support, some more stdlib: a64l, quad, tsearch
	sysconf support
	test for popen
	renamed fixcrcr.exe to fixtxt.exe
2000/04/18 (0.04b)
	additions: (arnd)
	err* stuff, __volatile__ macro fixed
	several minor CVS-Id changes
	msginfo structure renamed (clash with macro)
	removed timext.h
	_getPosixOS2Root cast fixed

2000/04/26 LIBREVISION 0.05
	additions: (arnd)
	perror, vis, man pages
	declare several arrays as pointers rather than []
	various conformance tests for FP and regex
	finally remove timext.h :-)
	set several constants from param.h which where hardcoded before
	init some local variables in regcomp to NULL
	empty getloadavg
	added some assembler math functions (i387math.S) from libgcc
	regexp debug function

2000/05/02 LIBREVISION 0.06
	additions: (arnd, Stefan Neis, Ralf Barkow)
	i387math.S fixes for acos/asin/exp from FreeBSD (neis)
	move fcntl.h after sys/types.h (barkow) and sys/stat.h (arnd)
	man pages: poll,bitstring,cat*,get*,killpg,math,hz,intro9
	i386 specific float.h, ieee.h->ieeefp.h
	nls support,killpg declaration, 
	IPC and SHM fixed,zlib integration
	mmap from BSD removed, nls messages for different languages
	sysconf fixes
	tests for db,shm
2000/05/05
	Math update (Cephes Math Lib), addition of SYS-facitilities.txt
	moved drand
	don't use gcc builtins
	handling of infinity
	test for floatingpoint funcs, consistency test
2000/05/08
	Math update s_tanh.c (neis)
2000/05/12
	Fix to stdio and unistd headers to make C++ happy (neis)

2000/05/15 LIBREVISION 0.07
	additions: ctype, libgen,
	docs for mmap, man5 pages
	tools: the ultimate editor (porting example), pmake
2000/05/17
	Minor update (arnd)
	additions: 4.3 compatible regex, realpath, nice
	more manpages
	HowTo pages for some OS/2 stuff
2000/05/31
	Update rev007b (arnd)
	fixcrcr.c: added check to detect binaries (\0 bytes in text)
	changed librevision format in dll.c
	added several EDM/2 articles and moved html files to docs/html
	some more manpages
	More stuff to FAQ
	renamed some fdlibm functions to __fdlibm
	__STRICT_ANSI__ removed from some defines
	register in prototypes removed
	itimer management improved
	added getsubopt,l64a,mergesort,bm,strlcpy,strmode,strsep
	non-Posix string extensions: strcasecmp etc.
2000/06/13
	math.h _LIB_VERSION var moved (neis)
2000/07/06
	minor changes of defs in math.h and limits.h for LONG LONG support
	(arnd)

Alpha #1 released

2000/07/24
	missing stuff for tools/pmake (arnd)
2000/07/25
	missing newline character added in some headers (amai)
	OpenBSD version of db added (arnd)
	misc fixes: headers, sources, docs (arnd)
2000/07/28
	<poll.h> added, minor updates for docs (amai)

2000/07/29 LIBREVISION 0.09
	lockf() added: not functional yet (see fcntl() in EMX) (arnd)
	timezone stuff added (arnd)
        Minor overall fixed (Makefiles, docs) (amai)

Alpha #2 released

2000/08/06
	Update rev009b
	time.h added, mmap fixes (arnd)
	removed #warnings from some headers - will make autoconf happy (arnd)
2000/08/08
	Update rev009c
	signgam added to libm (arnd)
	plain text version of FAQ removed in favour of HTML version (arnd)
	Makefile fixes (arnd, amai)
	[minor fixes to Makefiles and documentation are not always
	 logged here!]

2000/08/14 LIBREVISION 0.0.10
	misc fixes for libext (move some stuff in according subdirectories,
	dirent stuff, ...)   (arnd)
	Add big tree of utilities to tools/
2000/08/14 LIBREVISION 0.0.11
	few minor fixes for libExt (arnd)
	Modifications for Makefiles, make use of .mk files (arnd)
	"Missing" stuff in tools/ from rev 010 (arnd)
	working copy of rename(); tests for rename() and stat() (amai)
2000/08/15 LIBREVISION 0.0.11a
	few minor fixes for libExt; cleanup of directories (arnd)
	add data files to src/share (arnd)
	misc fixes for tools/ tree; added find(1) (arnd)
2000/08/18 LIBREVISION 0.0.11b
	Misc changes and additions; more files moved location (arnd)
	(In an early state this release was messed up in CVS!)
2000/08/19 LIBREVISION 0.0.12
	Misc stuff as usual. Mainly Makfiles this time,
	introduce two dlls instead of the single one
      change directory structures a bit (arnd)
2000/08/19 LIBREVISION 0.0.12a
2000/08/21 LIBREVISION 0.0.12b
2000/08/23 LIBREVISION 0.0.12c
	Dirent headers, glob* + misc stuff (arnd)
2000/08/26 LIBREVISION 0.0.13
	Build up basic directory structures (arnd)
      Add calendar and rmdir to tools/ as well as some manpages (arnd)
      Fixes/additions to libext
2000/09/09 LIBREVISION 0.0.14
      Add asa and printenv to tools/ as well as some manpages and 
      docs (arnd).
      Fixes/additions to libExt (arnd):
	1) New directory libext/libc/sys for modules that emulate BSD system 
	calls in a safer and more compatible way. Moved several ones.
	2) sys/mm.c enhanced and prepared for later split that will be 
	necessary for linking into a run-time DLL. New function 
	mprotect(). Now mmap() works with MAP_SHARED.
	3) sys/statx.c enhanced: New function statfs(). Several headers 
	adapted: sys/vfs.h, sys/statfs.h, sys/mount.h.
2000/10/11 LIBREVISION 0.0.14a (arnd)
	more changes of mm.c (MAP_SHARED/MAP_FIXED better supported)

tools/gencat/gencat.c
/*	$Id: gencat.c,v 1.2 2000/08/14 16:26:08 amai Exp $	*/

/*-
 * Copyright (c) 1996 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by J.T. Conklin.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgment:
 *        This product includes software developed by the NetBSD 
 *	  Foundation, Inc. and its contributors.
 * 4. Neither the name of The NetBSD Foundation nor the names of its 
 *    contributors may be used to endorse or promote products derived 
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/***********************************************************
Copyright 1990, by Alfalfa Software Incorporated, Cambridge, Massachusetts.

                        All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that Alfalfa's name not be used in
advertising or publicity pertaining to distribution of the software
without specific, written prior permission.

ALPHALPHA DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
ALPHALPHA BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

If you make any modifications, bugfixes or other changes to this software
we'd appreciate it if you could send a copy to us so we can keep things
up-to-date.  Many thanks.
				Kee Hinckley
				Alfalfa Software, Inc.
				267 Allston St., #3
				Cambridge, MA 02139  USA
				nazgul@alfalfa.com

******************************************************************/

#define _NLS_PRIVATE

/* ensure 8-bit cleanliness */
#define ISSPACE(c) (isascii(c) && isspace(c))

#include <sys/queue.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <nl_types.h>

extern void MCAddSet __P((int setId));
extern void MCDelSet __P((int setId));
extern void MCAddMsg __P((int msgId, const char *msg));
extern void MCDelMsg __P((int msgId));
extern void MCParse __P((int fd));
extern void MCWriteCat __P((int fd));

struct _msgT {
	long    msgId;
	char   *str;
        LIST_ENTRY(_msgT) entries;
};

struct _setT {
	long    setId;
        LIST_HEAD(msghead, _msgT) msghead;
        LIST_ENTRY(_setT) entries;
};

LIST_HEAD(sethead, _setT) sethead;
static struct _setT *curSet;

static char *curline = NULL;
static long lineno = 0;

void
usage(void)
{
	fprintf(stderr, "Use: gencat catfile msgfile ...\n");
	exit(1);
}

int main(int,char**);
int
main(argc, argv)
	int     argc;
	char   *argv[];
{
	int     ofd, ifd;
	char   *catfile = NULL;
	int     c;

	__progname = argv[0];
	_wildcard(&argc, &argv);


	while ((c = getopt(argc, argv, "")) != -1) {
		switch (c) {
		case '?':
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (argc < 2) {
		usage();
		/* NOTREACHED */
	}
	catfile = *argv++;

	for (; *argv; argv++) {
		if ((ifd = open(*argv, O_RDONLY)) < 0) {
			fprintf(stderr, "gencat: Unable to read %s\n", *argv);
			exit(1);
		}
		MCParse(ifd);
		close(ifd);
	}

	if ((ofd = open(catfile, O_WRONLY | O_TRUNC | O_CREAT, 0666)) < 0) {
		fprintf(stderr, "gencat: Unable to create a new %s.\n",
		    catfile);
		exit(1);
	}
	MCWriteCat(ofd);
	exit(0);
}

static void
warning(cptr, msg)
	char   *cptr;
	char   *msg;
{
	fprintf(stderr, "gencat: %s on line %ld\n", msg, lineno);
	fprintf(stderr, "%s\n", curline);
	if (cptr) {
		char   *tptr;
		for (tptr = curline; tptr < cptr; ++tptr)
			putc(' ', stderr);
		fprintf(stderr, "^\n");
	}
}

static void
error(cptr, msg)
	char   *cptr;
	char   *msg;
{
	warning(cptr, msg);
	exit(1);
}

static void
nomem()
{
	error(NULL, "out of memory");
}

static void *
xmalloc(len)
	size_t  len;
{
	void   *p;

	if ((p = malloc(len)) == NULL)
		nomem();
	return (p);
}

static void *
xrealloc(ptr, size)
	void   *ptr;
	size_t  size;
{
	if ((ptr = realloc(ptr, size)) == NULL)
		nomem();
	return (ptr);
}

static char *
xstrdup(str)
	char   *str;
{
	if ((str = strdup(str)) == NULL)
		nomem();
	return (str);
}

static char *
getline(fd)
	int     fd;
{
	static long curlen = BUFSIZ;
	static char buf[BUFSIZ], *bptr = buf, *bend = buf;
	char   *cptr, *cend;
	long    buflen;

	if (!curline) {
		curline = xmalloc(curlen);
	}
	++lineno;

	cptr = curline;
	cend = curline + curlen;
	for (;;) {
		for (; bptr < bend && cptr < cend; ++cptr, ++bptr) {
			if (*bptr == '\n') {
				*cptr = '\0';
				++bptr;
				return (curline);
			} else
				*cptr = *bptr;
		}
		if (bptr == bend) {
			buflen = read(fd, buf, BUFSIZ);
			if (buflen <= 0) {
				if (cptr > curline) {
					*cptr = '\0';
					return (curline);
				}
				return (NULL);
			}
			bend = buf + buflen;
			bptr = buf;
		}
		if (cptr == cend) {
			cptr = curline = xrealloc(curline, curlen *= 2);
			cend = curline + curlen;
		}
	}
}

static char *
wskip(cptr)
	char   *cptr;
{
	if (!*cptr || !ISSPACE(*cptr)) {
		warning(cptr, "expected a space");
		return (cptr);
	}
	while (*cptr && ISSPACE(*cptr))
		++cptr;
	return (cptr);
}

static char *
cskip(cptr)
	char   *cptr;
{
	if (!*cptr || ISSPACE(*cptr)) {
		warning(cptr, "wasn't expecting a space");
		return (cptr);
	}
	while (*cptr && !ISSPACE(*cptr))
		++cptr;
	return (cptr);
}

static char *
getmsg(fd, cptr, quote)
	int     fd;
	char   *cptr;
	char    quote;
{
	static char *msg = NULL;
	static long msglen = 0;
	long    clen, i;
	char   *tptr;

	if (quote && *cptr == quote) {
		++cptr;
	};

	clen = strlen(cptr) + 1;
	if (clen > msglen) {
		if (msglen)
			msg = xrealloc(msg, clen);
		else
			msg = xmalloc(clen);
		msglen = clen;
	}
	tptr = msg;

	while (*cptr) {
		if (quote && *cptr == quote) {
			char   *tmp;
			tmp = cptr + 1;
			if (*tmp && (!ISSPACE(*tmp) || *wskip(tmp))) {
				warning(cptr, "unexpected quote character, ignoring");
				*tptr++ = *cptr++;
			} else {
				*cptr = '\0';
			}
		} else
			if (*cptr == '\\') {
				++cptr;
				switch (*cptr) {
				case '\0':
					cptr = getline(fd);
					if (!cptr)
						error(NULL, "premature end of file");
					msglen += strlen(cptr);
					i = tptr - msg;
					msg = xrealloc(msg, msglen);
					tptr = msg + i;
					break;
				case 'n':
					*tptr++ = '\n';
					++cptr;
					break;
				case 't':
					*tptr++ = '\t';
					++cptr;
					break;
				case 'v':
					*tptr++ = '\v';
					++cptr;
					break;
				case 'b':
					*tptr++ = '\b';
					++cptr;
					break;
				case 'r':
					*tptr++ = '\r';
					++cptr;
					break;
				case 'f':
					*tptr++ = '\f';
					++cptr;
					break;
				case '\\':
					*tptr++ = '\\';
					++cptr;
					break;
				case '"': 
				case '\'':
					/* 
					 * While it isn't necessary to
					 * escape ' and ", let's accept
					 * them escaped and not complain.
					 * (XPG4 states that '\' should be
					 * ignored when not used in a
					 * valid escape sequence)
					 */
					*tptr++ = '"';
					++cptr;
					break;
				default:
					if (isdigit(*cptr)) {
						*tptr = 0;
						for (i = 0; i < 3; ++i) {
							if (!isdigit(*cptr))
								break;
							if (*cptr > '7')
								warning(cptr, "octal number greater than 7?!");
							*tptr *= 8;
							*tptr += (*cptr - '0');
							++cptr;
						}
					} else {
						warning(cptr, "unrecognized escape sequence; ignoring escape character");
					}
				}
			} else {
				*tptr++ = *cptr++;
			}
	}
	*tptr = '\0';
	return (msg);
}

void
MCParse(fd)
	int     fd;
{
	char   *cptr, *str;
	int     setid, msgid = 0;
	char    quote = 0;

	/* XXX: init sethead? */

	while ((cptr = getline(fd))) {
		if (*cptr == '$') {
			++cptr;
			if (strncmp(cptr, "set", 3) == 0) {
				cptr += 3;
				cptr = wskip(cptr);
				setid = atoi(cptr);
				MCAddSet(setid);
				msgid = 0;
			} else if (strncmp(cptr, "delset", 6) == 0) {
				cptr += 6;
				cptr = wskip(cptr);
				setid = atoi(cptr);
				MCDelSet(setid);
			} else if (strncmp(cptr, "quote", 5) == 0) {
				cptr += 5;
				if (!*cptr)
					quote = 0;
				else {
					cptr = wskip(cptr);
					if (!*cptr)
						quote = 0;
					else
						quote = *cptr;
				}
			} else if (ISSPACE(*cptr)) {
				;
			} else {
				if (*cptr) {
					cptr = wskip(cptr);
					if (*cptr)
						warning(cptr, "unrecognized line");
				}
			}
		} else {
			if (isdigit(*cptr)) {
				msgid = atoi(cptr);
				cptr = cskip(cptr);
				cptr = wskip(cptr);
				/* if (*cptr) ++cptr; */
			}
			if (!*cptr)
				MCDelMsg(msgid);
			else {
				str = getmsg(fd, cptr, quote);
				MCAddMsg(msgid, str);
			}
		}
	}
}

/*
 * Write message catalog.
 *
 * The message catalog is first converted from its internal to its
 * external representation in a chunk of memory allocated for this
 * purpose.  Then the completed catalog is written.  This approach
 * avoids additional housekeeping variables and/or a lot of seeks
 * that would otherwise be required.
 */
void
MCWriteCat(fd)
	int     fd;
{
	int     nsets;		/* number of sets */
	int     nmsgs;		/* number of msgs */
	int     string_size;	/* total size of string pool */
	int     msgcat_size;	/* total size of message catalog */
	void   *msgcat;		/* message catalog data */
	struct _nls_cat_hdr *cat_hdr;
	struct _nls_set_hdr *set_hdr;
	struct _nls_msg_hdr *msg_hdr;
	char   *strings;
	struct _setT *set;
	struct _msgT *msg;
	int     msg_index;
	int     msg_offset;

	/* determine number of sets, number of messages, and size of the
	 * string pool */
	nsets = 0;
	nmsgs = 0;
	string_size = 0;

	for (set = sethead.lh_first; set != NULL;
	    set = set->entries.le_next) {
		nsets++;

		for (msg = set->msghead.lh_first; msg != NULL;
		    msg = msg->entries.le_next) {
			nmsgs++;
			string_size += strlen(msg->str) + 1;
		}
	}

#ifdef DEBUG
	printf("number of sets: %d\n", nsets);
	printf("number of msgs: %d\n", nmsgs);
	printf("string pool size: %d\n", string_size);
#endif

	/* determine size and then allocate buffer for constructing external
	 * message catalog representation */
	msgcat_size = sizeof(struct _nls_cat_hdr)
	    + (nsets * sizeof(struct _nls_set_hdr))
	    + (nmsgs * sizeof(struct _nls_msg_hdr))
	    + string_size;

	msgcat = xmalloc(msgcat_size);
	memset(msgcat, '\0', msgcat_size);

	/* fill in msg catalog header */
	cat_hdr = (struct _nls_cat_hdr *) msgcat;
	cat_hdr->__magic = htonl(_NLS_MAGIC);
	cat_hdr->__nsets = htonl(nsets);
	cat_hdr->__mem = htonl(msgcat_size - sizeof(struct _nls_cat_hdr));
	cat_hdr->__msg_hdr_offset =
	    htonl(nsets * sizeof(struct _nls_set_hdr));
	cat_hdr->__msg_txt_offset =
	    htonl(nsets * sizeof(struct _nls_set_hdr) +
	    nmsgs * sizeof(struct _nls_msg_hdr));

	/* compute offsets for set & msg header tables and string pool */
	set_hdr = (struct _nls_set_hdr *) ((char *) msgcat +
	    sizeof(struct _nls_cat_hdr));
	msg_hdr = (struct _nls_msg_hdr *) ((char *) msgcat +
	    sizeof(struct _nls_cat_hdr) +
	    nsets * sizeof(struct _nls_set_hdr));
	strings = (char *) msgcat +
	    sizeof(struct _nls_cat_hdr) +
	    nsets * sizeof(struct _nls_set_hdr) +
	    nmsgs * sizeof(struct _nls_msg_hdr);

	msg_index = 0;
	msg_offset = 0;
	for (set = sethead.lh_first; set != NULL;
	    set = set->entries.le_next) {

		nmsgs = 0;
		for (msg = set->msghead.lh_first; msg != NULL;
		    msg = msg->entries.le_next) {
			int     msg_len = strlen(msg->str) + 1;

			msg_hdr->__msgno = htonl(msg->msgId);
			msg_hdr->__msglen = htonl(msg_len);
			msg_hdr->__offset = htonl(msg_offset);

			memcpy(strings, msg->str, msg_len);
			strings += msg_len;
			msg_offset += msg_len;

			nmsgs++;
			msg_hdr++;
		}

		set_hdr->__setno = htonl(set->setId);
		set_hdr->__nmsgs = htonl(nmsgs);
		set_hdr->__index = htonl(msg_index);
		msg_index += nmsgs;
		set_hdr++;
	}

	/* write out catalog.  XXX: should this be done in small chunks? */
	write(fd, msgcat, msgcat_size);
}

void
MCAddSet(setId)
	int     setId;
{
	struct _setT *p, *q;

	if (setId <= 0) {
		error(NULL, "setId's must be greater than zero");
		/* NOTREACHED */
	}
#if 0
	/* XXX */
	if (setId > NL_SETMAX) {
		error(NULL, "setId %d exceeds limit (%d)");
		/* NOTREACHED */
	}
#endif

	p = sethead.lh_first;
	q = NULL;
	for (; p != NULL && p->setId < setId; q = p, p = p->entries.le_next);

	if (p && p->setId == setId) {
		;
	} else {
		p = xmalloc(sizeof(struct _setT));
		memset(p, '\0', sizeof(struct _setT));
		LIST_INIT(&p->msghead);

		p->setId = setId;

		if (q == NULL) {
			LIST_INSERT_HEAD(&sethead, p, entries);
		} else {
			LIST_INSERT_AFTER(q, p, entries);
		}
	}

	curSet = p;
}

void
MCAddMsg(msgId, str)
	int     msgId;
	const char *str;
{
	struct _msgT *p, *q;

	if (!curSet)
		error(NULL, "can't specify a message when no set exists");

	if (msgId <= 0) {
		error(NULL, "msgId's must be greater than zero");
		/* NOTREACHED */
	}
#if 0
	/* XXX */
	if (msgId > NL_SETMAX) {
		error(NULL, "msgId %d exceeds limit (%d)");
		/* NOTREACHED */
	}
#endif

	p = curSet->msghead.lh_first;
	q = NULL;
	for (; p != NULL && p->msgId < msgId; q = p, p = p->entries.le_next);

	if (p && p->msgId == msgId) {
		free(p->str);
	} else {
		p = xmalloc(sizeof(struct _msgT));
		memset(p, '\0', sizeof(struct _msgT));

		if (q == NULL) {
			LIST_INSERT_HEAD(&curSet->msghead, p, entries);
		} else {
			LIST_INSERT_AFTER(q, p, entries);
		}
	}

	p->msgId = msgId;
	p->str = xstrdup(str);
}

void
MCDelSet(setId)
	int     setId;
{
	struct _setT *set;
	struct _msgT *msg;

	set = sethead.lh_first;
	for (; set != NULL && set->setId < setId; set = set->entries.le_next);

	if (set && set->setId == setId) {

		msg = set->msghead.lh_first;
		while (msg) {
			free(msg->str);
			LIST_REMOVE(msg, entries);
		}

		LIST_REMOVE(set, entries);
		return;
	}
	warning(NULL, "specified set doesn't exist");
}

void
MCDelMsg(msgId)
	int     msgId;
{
	struct _msgT *msg;

	if (!curSet)
		error(NULL, "you can't delete a message before defining the set");

	msg = curSet->msghead.lh_first;
	for (; msg != NULL && msg->msgId < msgId; msg = msg->entries.le_next);

	if (msg && msg->msgId == msgId) {
		free(msg->str);
		LIST_REMOVE(msg, entries);
		return;
	}
	warning(NULL, "specified msg doesn't exist");
}
